CI: treat `XDeclaredButNotUsed` as an error

zerbina opened the pull request:
## Summary

* fix all `XDeclaredButNotUsed` warnings in the code bases of the
  compiler and tooling
* enable the hint for the compiler and associated tooling
* promote the hint to an error when strict mode (`-d:nimStrictMode`) is
  enabled

The aim is to prevent obsolete code from being merged or being left
behind after a cleanup/refactor.

## Details

The warnings are fixed through a mixture of approaches:
* unused routines and types that were unused for a long time
  and where it's unlikely that they'll become used again are removed
* unused locals with side-effect-less initializer are removed as a
  whole
* unused locals with side-effect-ful initializer expressions are
  replaced with `discard` statements
* symbols injected by templates are marked as `.used`
* currently unused routines and locals meant for debugging are kept and
  marked as `.used`
* some unused locals are not removed but instead made use of

A few places are of special note:
* the unused `a0` local in `semTemplBody` was assigned the result of
  a redundant `semTemplBody` call. The call does have side-effects, but
  thanks to in-place AST modifications the duplicate processing didn't
  cause problems
* the `assign` procedure in the `checked_ast` module was meant from the
  start to be exported
* there are multiple routines in the `suggest` module that are unused
  when building the compiler itself. `XDeclaredButNotUsed` hints are
  now disabled for the module when not building with the `nimsuggest`
  define

Finally, disabling the `XDeclaredButNotUsed` is removed from all
configuration files and the hint is promoted to an error when
`nimStrictMode` is enabled.

zerbina committed:
mark symbols with `.used` where appropriate

Symbols injected by templates, useful debug routines, and symbols where
it's likely that they'll be used again in the future are marked as
`.used` in order to silence the hint.

zerbina committed:
remove unused symbols

The removals are pure removals of unused routines, types, and
locals.

zerbina committed:
checked_ast: export `assign`

The procedure was wrongfully not marked as exported at its inception.

zerbina committed:
fix unused local symbol warnings

zerbina committed:
semtypes: report annotation key errors

This is not a fix, as the error was still reported as part of the
`semOverloadedCall` call.

zerbina committed:
semtempl: remove redundant processing

The destination expression of an assignment-to-bracket-expr was
processed twice. While the processing is effectful, the in-place
modifications of the AST meant that the identifiers weren't processed
twice (which is what would have caused problems).

zerbina committed:
suggest: disable `XDeclaredButNotUsed` hints

Ideally, all unused routines would be within `when defined(nimsuggest)`
sections, but since this is not the case and changing it would only
clutter the Git history, the hint is disabled within the module when
building without the `nimsuggest` define.

zerbina committed:
config: don't disable `XDeclaredButNotUsed` hint

zerbina committed:
config: `XDeclaredButNotUsed` is an error in strict mode

saem commented on compiler/sem/nilcheck.nim:
```diff
@@ -358,7 +351,8 @@ proc namedSetsDebugInfo(ctx: NilCheckerContext, map: NilMap): string =
 
-proc namedMapAndSetsDebugInfo(ctx: NilCheckerContext, map: NilMap): string =
+proc namedMapAndSetsDebugInfo(ctx: NilCheckerContext,
+                              map: NilMap): string {.used.} =
```
Non-blocker: should debug routines be marked as `{.deprecated: "do not use in final build".}`, similar to `astrepr`?

zerbina replied:
Hm, yeah, that sounds like a good idea -- I'll change it.

saem reviewed (approved):
One non-blocking item, otherwise this is good to merge. It'll be nice having tools take on this workload.

zerbina committed:
nilcheck: add the `.debug` pragma

In addition, to marking a procedure as used, it also tags it as
deprecated, leading to a warning when using the procedure

zerbina commented:
/merge
